panda/src/putil/pointerData.h

```diff
@@ -29,40 +29,85 @@ enum class PointerType {
   stylus,
   eraser,
 };
+
+/**
+ * The various states that the pointer can be in during a full sequence.
+ */
+enum class PointerPhase {
+  began, // made contact with the screen during the last frame
+  moved, // continued interaction with the screen
+  ended, // contact ended normally by lifting the pointer or leaving the window.
+  cancelled, // contact was interrupted
+};
```

Comment:
I'm feeling a little uneasy about encoding a state transition in the form of a state that is held for one frame, but I'll have to dig more into the code to see why it is done this way and whether it could be done differently.